dojo/tools/csaf/parser.py

```diff
@@ -0,0 +1,163 @@
+import json
+import re
+
+from os.path import join
+
+from django.urls import reverse
+from dojo.models import Finding
+from dojo.notifications.helper import create_notification
+from jsonschema import ValidationError, validate
+
+
+class CsafParser(object):
+    """CSAF Scanner JSON Report"""
+
+    def get_scan_types(self):
+        return ["CSAF Scan"]
+
+    def get_label_for_scan_types(self, scan_type):
+        return "CSAF Scan"
+
+    def get_description_for_scan_types(self, scan_type):
+        return "CSAF JSON report format"
+
+    # file: django.core.files.uploadedfile.TemporaryUploadedFile
+    def get_findings(self, file, dd_test):
+        if file is None:
+            raise Exception("No report file provided")
+
+        csaf = json.load(file)
+
+        # Raises on validation error.
+        csaf_validate(dd_test, csaf)
+
+        return csaf_import(dd_test, file, csaf)
+
+
+def csaf_validate(dd_test, csaf):
+    # https://github.com/oasis-tcs/csaf/raw/master/csaf_2.0/json_schema/csaf_json_schema.json
+    schema_path = join('dojo', 'tools', 'csaf', 'csaf_json_schema.json')
+    with open(schema_path, encoding="utf8") as csaf_schema_file:
+        csaf_schema_json = json.load(csaf_schema_file)
+
+    try:
+        validate(csaf, schema=csaf_schema_json)
+    except ValidationError as e:
+        # `validate` raises on validation errors with a huge error
+        # message complaining about the exact position of the validation
+        # error.
+        # The first line of said message is a summary, so we just
+        # re`raise` using this line.
+        reason = repr(e)
+
+        create_notification(event='other',
+                            title=reason,
+                            description=e,
+                            test=dd_test,
+                            engagement=dd_test.engagement,
+                            product=dd_test.engagement.product,
+                            url=reverse('alerts'))
+
+        raise ValidationError(reason)
+
+
+def csaf_import(dd_test, file, json):
+    document = json.get('document', {})
+    vulnerabilities = json.get('vulnerabilities', {})
+    product_tree = json.get('product_tree', {})
+
+    engagement_title = document.get('title', '(No Title)')
+
+    if hasattr(dd_test, 'engagement'):
+        dd_test.engagement.name = engagement_title
```

Comment:
this is really weird.
the title of the engagement will change for every import.
Not what DefectDojo engagement are for.
If you want to customize the Test title, you can do it by using the other parser interface (see SARIF/CycloneDX implementation)